File: mac/Sources/FocusKit/Models/Grain.swift
import Foundation

public enum Grain: String, CaseIterable, Identifiable {
    case day, week, month, year

    public var id: String { rawValue }

    public var label: String {
        switch self {
        case .day: return "Daily"
        case .week: return "Weekly"
        case .month: return "Monthly"
        case .year: return "Yearly"
        }
    }

    public var component: Calendar.Component {
        switch self {
        case .day: return .day
        case .week: return .weekOfYear
        case .month: return .month
        case .year: return .year
        }
    }

    /// How many periods the chart looks back over.
    public var span: Int {
        switch self {
        case .day: return 14
        case .week: return 12
        case .month: return 12
        case .year: return 5
        }
    }

    /// Formatters are kept rather than made on demand: building one costs more
    /// than formatting with it, and the chart formats a label per bar.
    private static let cache = FormatterCache()

    public func title(for date: Date, calendar: Calendar) -> String {
        let formatter = Grain.cache.formatter(for: "\(rawValue).title", calendar: calendar)
        switch self {
        case .day: formatter.dateFormat = "EEE d MMM"
        case .week: formatter.dateFormat = "'w/c' d MMM"
        case .month: formatter.dateFormat = "MMMM yyyy"
        case .year: formatter.dateFormat = "yyyy"
        }
        return formatter.string(from: date)
    }

    public func shortTitle(for date: Date, calendar: Calendar) -> String {
        let formatter = Grain.cache.formatter(for: "\(rawValue).short", calendar: calendar)
        switch self {
        case .day: formatter.dateFormat = "EEEEE"
        case .week: formatter.dateFormat = "d MMM"
        case .month: formatter.dateFormat = "MMM"
        case .year: formatter.dateFormat = "yyyy"
        }
        return formatter.string(from: date)
    }
}

/// One date formatter per format string, made once and reused.
final class FormatterCache: @unchecked Sendable {
    private var formatters: [String: DateFormatter] = [:]
    private let lock = NSLock()

    func formatter(for key: String, calendar: Calendar) -> DateFormatter {
        lock.lock()
        defer { lock.unlock() }
        if let existing = formatters[key] { return existing }
        let formatter = DateFormatter()
        formatter.calendar = calendar
        formatters[key] = formatter
        return formatter
    }
}

File: windows/Focus.Core/Models/Grain.cs
namespace Focus.Core;

public enum Grain { Day, Week, Month, Year }

public static class Grains
{
    public static readonly Grain[] All = { Grain.Day, Grain.Week, Grain.Month, Grain.Year };

    public static string Label(this Grain grain) => grain switch
    {
        Grain.Day => "Daily",
        Grain.Week => "Weekly",
        Grain.Month => "Monthly",
        _ => "Yearly"
    };

    /// How many periods the chart looks back over.
    public static int Span(this Grain grain) => grain switch
    {
        Grain.Day => 14,
        Grain.Week => 12,
        Grain.Month => 12,
        _ => 5
    };

    public static string Title(this Grain grain, DateTime date) => grain switch
    {
        Grain.Day => date.ToString("ddd d MMM"),
        Grain.Week => "w/c " + date.ToString("d MMM"),
        Grain.Month => date.ToString("MMMM yyyy"),
        _ => date.ToString("yyyy")
    };

    public static string ShortTitle(this Grain grain, DateTime date) => grain switch
    {
        Grain.Day => date.ToString("ddd").Substring(0, 1),
        Grain.Week => date.ToString("d MMM"),
        Grain.Month => date.ToString("MMM"),
        _ => date.ToString("yyyy")
    };
}
